### ISSUE TITLE
3.2 hangs forever reading from stdin on our Jenkins/Docker CI

### ISSUE THREAD

**Issue by stephank** [squizlabs/PHP_CodeSniffer#1788]
Hi! There's a bit of a complicated issue going on between the stdin detection changes and the way Jenkins with the Docker pipeline plugin runs things.

We basically have a project with phpcs as a dev dependency, and a Dockerfile from which we build the project. Then we run phpcs inside that image. In our Jenkinsfile that looks like:

```groovy
let image = docker.build('foobar')
image.inside {
  sh '''#!/bin/sh -xe
    ./vendor/bin/phpcs -p --standard=PSR2 --extensions=php src/
  '''
}
```

Even though we don't specify stdin on the commandline, phpcs appears to hang reading from stdin, which blocks indefinitely. If I tack on `< /dev/null` at the end of the phpcs command, things work again. 3.1.1 also works fine.

I'll try to get some more info. It should be possible to reproduce this with Docker alone.

**Comment by stephank** [comment]
It appears I cannot run phpcs in Docker on my local macOS with Docker for Mac setup either. I tried:

```sh
docker run --rm -it -v $PWD:/src -w /src php ./vendor/bin/phpcs -p --standard=PSR2 --extensions=php src/
```

Or just starting a shell in a PHP container, and running it from there:

```sh
docker run --rm -it -v $PWD:/src -w /src php /bin/sh
# ./vendor/bin/phpcs -p --standard=PSR2 --extensions=php src/
```

But strangely, the `< /dev/null` trick doesn't work here, whereas it does in Jenkins.

It does respond to `--help` and also shows usage when you run it simply without args.

**Comment by kissifrot** [comment]
It seems it also hangs when used from PhpStorm:
https://youtrack.jetbrains.com/issue/WI-39190
https://stackoverflow.com/questions/47797797/error-with-php-code-sniffer-and-phpstorm

**Comment by gsherwood** [comment]
Any chance you could test master? I've committed a fix for the PHPStorm issue already (just needs some more testing) and it would be good to see if that's the same issue. If that fix checks out for you as well, I can do a 3.2.1 release to resolve it.

**Comment by eclipxe13** [comment]
I tested locally using PHPStorm and it works fine! thanks @gsherwood

Running on Debian GNU linux Buster (testing)
Phpstorm 2017.3
PHP_CodeSniffer version 3.2.1 (at commit 6172062f0)

**Comment by gsherwood** [comment]
@eclipxe13 Thanks for testing.

**Comment by KacerCZ** [comment]
I have same problem with NetBeans IDE (nightly build 201710240002) and PHP_CodeSniffer 3.2.0.
I tested PHP_CodeSniffer commit f130348 and it works as expected.

**Comment by gsherwood** [comment]
@KacerCZ Thanks for testing it out. I didn't know there was an issue with the NetBeans plugin as well.

**Comment by toonvanacker** [comment]
We're also getting this issue when running PHPCS during our Bamboo build process. Our fix was to go back to 3.1.1.

**Comment by stephank** [comment]
I think whatever problem I was having trying to reproduce the issue in just Docker (without Jenkins) was a red herring. Docker for Mac just has a really slow filesystem, and the directory scan was taking ages.

I've managed to test master in a real CI run, and confirm the issue is fixed here too. :)

**Comment by Pierozi** [comment]
Hello there, I confirm the same error at our platform [ContinuousPHP](http://continuousphp.com/)
This is no related to the fix of #994 
The `stream_select` will stay in blocked mode forever

**Comment by Pierozi** [comment]
Previously the `checkStdin` was at false, now `defined('STDIN')` is set to `true`

**Comment by gsherwood** [comment]
Posting here instead of the PR so the discussion is centralised:

@Pierozi The stream was set to blocking on purpose during that change by @arnested - see comment: https://github.com/squizlabs/PHP_CodeSniffer/issues/1472#issuecomment-302332954

Are you able to provide a bit more info about what you are doing? You said that `$checkStdin` used to be false for you, which means you might be passing CLI args directly to the Config constructor. Is this the case?

I had a similar problem with a script that was wrapping and calling PHPCS directly, but I found it was caused by me opening STDIN before calling PHPCS (and not closing it) so it was waiting for feof. Maybe you are doing something similar?

Either way, there is a timeout in the master now (as there was previously). Have you tried this latest code to see if it works for you?

**Comment by Pierozi** [comment]
@gsherwood you absolute right, I missed your [commit](https://github.com/squizlabs/PHP_CodeSniffer/commit/6172062f09eccea2c851cbc11daad137ebe52f98) because I rush on PR and it was not there.

I confirm your timeout worked well to avoid the eof issue.
All will be back to normal on `3.2.1`

Many thanks for your involvement.

**Comment by gsherwood** [comment]
@Pierozi Thanks a lot for testing that for me.

**Comment by gsherwood** [comment]
Based on the testing done, I'm marking this as fixed in 3.2.1
